### PR TITLE
verbs RDM. fix of debug build + refactoring send/recv part + rma read/write/inject

### PR DESCRIPTION
**Pull request by evgeny-leksikov**
Reopen #1604 - excluded merge commit

1) Commit #3c27306 fixes debug build
Others are nice to have
2) Refactoring send/recv paths + request structure
3) Implementation of rma read/write/inject

### CODE REVIEW

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 1]
```diff
@@ -75,6 +75,42 @@
...
         ((void*)(uintptr_t)(value & (~(FI_IBV_RDM_SERVICE_WR_MASK))))
 
+// Send/Recv counters control
+
+#define FI_IBV_RDM_TAGGED_INC_SEND_COUNTERS(_connection, _ep, _send_flags) \
+do {                                                                	\
+	(_connection)->sends_outgoing++;                                \
+	(_ep)->pend_send++;                                             \
+	(_ep)->total_outgoing_send++;                                   \
+	(_send_flags) |= IBV_SEND_SIGNALED;                             \
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Based on the function name, modifying the send_flags looks like a side effect.

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 2]
```diff
@@ -90,46 +126,65 @@ struct fi_ibv_rdm_tagged_rndv_header {
...
+	/* Accessors and match info */
+
 	struct fi_ibv_mem_pool_entry mpe;
+	/* Request can be an element of only one queue at the moment */
+	struct dlist_entry queue_entry;
+
+	struct {
+		enum fi_ibv_rdm_tagged_request_eager_state eager;
+		enum fi_ibv_rdm_tagged_request_rndv_state rndv;
+	} state;
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Should this be a union?

**Comment by shefty** [comment]
I had a couple of comments, which can either be ignored if they don't apply or addressed in separate patches.

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 1 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: I'm agree with naming but decreasing occurs by handling of local completion request, so signaled flag is mandatory. I added to my todo list to rename this macro like "INC_SIG_SEND_COUNTERS"

File: prov/verbs/src/ep_rdm/verbs_rdm.h (hunk 2 above)
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: No, these two fields are dimensions of 3-d state machine handlers array, third one is event

**Pull request merged**